spirv-fuzz: Swap positions of two functions in a module

sliu-UIUC opened the pull request:
This PR resolves #4026 by adding a new transformation class `transformation_swap_two_functions` that swaps the syntactic position of two functions in the module.

Contribution:
- Implemented transformation class `transformation_swap_two_functions`. The transformation class assumes that the two functions being swapped cannot be the same one (i.e., have the same id). `IsApplicable` checks function with the two given ids can be found within a module. The `Apply` method finds two function iterators with given id by calling `GetFunctionIterator()`, assert that both can be found, and swap the two function iterators. 

- Implemented `GetFunctionIteartor()` function in `fuzzerutil` that loops through all function ids in the module and return the function iterator that matches the given function id. 

- Added fuzzer pass `fuzzer_pass_swap_functions` that applies the new transformation class. 

- Added unit testing for the new transformation class in `test/fuzz`. It tests scenarios where either one of the ids is not attainable, two function ids are the same, as well as valid input that will result in successful swapping of the given two function ids. It also checks whether the correct SPIRV assembly string is acquired after applying the transformation.

paulthomson commented on source/fuzz/transformation_swap_two_functions.h:
```diff
@@ -0,0 +1,58 @@
+      protobufs::TransformationSwapTwoFunctions message);
+
+  TransformationSwapTwoFunctions(uint32_t function_id1, uint32_t function_id2);
+  // - |function_id1| is a valid non-entry-point OpFunction instruction
```
Add an empty line above the comment.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,250 @@
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  // Function should not swap with itself
+
+  
+  // Permutation has invalid values 2
```
These comments don't seem correct.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,250 @@
+                   .IsApplicable(context.get(), transformation_context),
+               "Two functions cannot be the same.");
+
+//   // Function with id 29 does not exist. 
```
Remove extra `//`.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,250 @@
+               "Two functions cannot be the same.");
+
+//   // Function with id 29 does not exist. 
+  ASSERT_DEATH( TransformationSwapTwoFunctions(10,29)
+                   .IsApplicable(context.get(), transformation_context),
+               "Function 2 is not in range.");
```
Once you have addressed my comments in IsApplicable, this can be changed to ASSERT_FALSE

sliu-UIUC replied:
Is there a preference for using ``ASSERT_FALSE`` over ``ASSERT_DEATH`` whenever possible? What are some of the situations where ``ASSERT_DEATH``/``ASSERT_FALSE`` might be preferred

paulthomson replied:
ASSERT_FALSE(X, Y) asserts that the expression X evaluates to false. If X does not evaluate to false, the test fails.

ASSERT_DEATH(X, Y) does something completely different. In fact, it is a bit magical. It calls `fork()` to create a child process to evaluate the expression X, and asserts that the process **crashes** (exits with an error status) with a given error message Y. If the error message does not contain Y, then the test fails.

In the above example, I am basically saying that `IsApplicable` should NOT crash when given a function ID that is not in the SPIR-V module. This is a situation that will often happen when using `spirv-fuzz` to shrink a test case by replaying a sublist of transformations. We want to handle this case by returning false so that the transformation is skipped. This is very likely to happen because the shrinker might be given a list of transformations like:

`[ TransformationAddFunction(2, ...), TransformationSwapTwoFunctions(2, 3) ]`

And the shrinker might try to apply a sublist of the transformations:

`[ TransformationSwapTwoFunctions(2, 3) ]`

And in this case, function 2 won't exist, so `TransformationSwapTwoFunctions.IsApplicable()` will return false, and so the transformation won't be applied, which is fine.

sliu-UIUC replied:
Fixed.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,250 @@
+               OpReturnValue %35
+               OpFunctionEnd
+  )";
+  // Final check to make sure the serious transformation above is correct. 
```
I guess by "serious" you mean "applied".

sliu-UIUC replied:
Meant "series of"... a combination of typo and autocomplete can be serious

sliu-UIUC replied:
Fixed.

paulthomson commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,63 @@
+        function_ids.emplace_back(function.result_id());
+    }
+    
+    int id_size = function_ids.size(); 
```
For me, this lines gives an error (warning as error):

```
error: implicit conversion loses integer precision: 'std::vector::size_type' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
```
I suggest using size_t or explicitly casting.

sliu-UIUC replied:
Fixed.

paulthomson commented on source/fuzz/CMakeLists.txt:
```diff
@@ -311,6 +313,7 @@ if(SPIRV_BUILD_FUZZER)
         fuzzer_pass_split_blocks.cpp
         fuzzer_pass_swap_commutable_operands.cpp
         fuzzer_pass_swap_conditional_branch_operands.cpp
+        fuzzer_pass_swap_functions.cpp
```
I get a compile error (warning) because the transformation is not handled in transformation.cpp:

```
/data/git/SPIRV-Tools/source/fuzz/transformation.cpp:114:11: error: enumeration value 'kSwapTwoFunctions' not handled in switch [-Werror,-Wswitch]
  switch (message.transformation_case()) {

```

sliu-UIUC replied:
Added `kSwapTwoFunctions` to `transformation.cpp` and is now fixed.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,250 @@
+  ASSERT_TRUE(
+      swap_test5.IsApplicable(context.get(), transformation_context));
+  
+  std::string after_transformation = R"(
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
```
The entire shader text is identical to the above shader text. Thus, the test is wrong. You need to swap function 10 and 13 in the shader text; that is what we expect to see.

Also, you did not `Apply` the `swap_test5` transformation; you can do this by calling: `ApplyAndCheckFreshIds(swap_test5, context.get(), &transformation_context);` 

However, even after doing this locally, the module does not seem to change. Thus, the test will fail. But it would still be good to fix the test first (so that it fails).

sliu-UIUC replied:
This is strange because it seems ``ASSERT_TRUE`` applied the change, since the change matched my new shader script with is assembled after I switched the functions.

sliu-UIUC replied:
Fixed unit test and should be working now.

paulthomson reviewed (commented):
Getting closer, but there are still a few issues. Please see comments.

paulthomson commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,63 @@
+    
+    int id_size = function_ids.size(); 
+    // We iterate through every combination of id i & j where i!=j.
+    for(int i = 0; i<id_size-1; ++i) {
+       for(int j = i+1; j<id_size; ++j) {  
```
The code is also still not formatted correctly, and thus it cannot be merged because the format check fails. You need to use clang-format-5. The README contains instructions for downloading it.

sliu-UIUC replied:
Clang-format done.

paulthomson commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,76 @@
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // Found the two functions in ir_context and swap their position. 
+
+  opt::Function* func1_ptr = ir_context->GetFunction(message_.function_id1());
+  opt::Function* func2_ptr = ir_context->GetFunction(message_.function_id2());
+
+  assert( func1_ptr!=nullptr  && "ERROR: Function 1 was not found with the given id."); 
+  assert( func2_ptr!=nullptr  && "ERROR: Function 2 was not found with the given id.");
+  assert( &func1_ptr != &func2_ptr && "ERROR: Two functions cannot be the same.");
+
+  std::swap(func1_ptr, func2_ptr); 
```
This does not quite work; it just swaps the local variables.

Even though we suggested using GetFunction, it turns out that swapping two functions is more easily done using iterators. So please could you obtain two iterators (e.g. `function_1` and `function_2`), and then do:

```
std::iter_swap(function_1.Get(), function_2.Get());
```

Or:

```
std::swap(*function_1.Get(), *function_2.Get());
```


The `.Get()` method returns the underlying iterator from the `UptrVectorIterator`, which is an iterator over `vector<unique_ptr<...>>`. And the `unique_ptr` type supports `swap`.

sliu-UIUC replied:
Fixed. Now function iterator is attained by calling `GetFunctionIterator(id)` from `fuzzerutil` and use `std::iter_swap` to perform function swaps.

sliu-UIUC commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,63 @@
+         // We do a swap between functions and break if such swap cannot be performed.
+         TransformationSwapTwoFunctions transformation(function_ids[i], function_ids[j]); 
+         if(!MaybeApplyTransformation(transformation)) {
+           break; 
```
This is where I see other codes have, should I simply remove it?

sliu-UIUC replied:
Removed.

paulthomson commented on kokoro/check-format/build.sh:
```diff
@@ -35,7 +35,8 @@ git clone https://github.com/google/googletest          external/googletest
 git clone --depth=1 https://github.com/google/effcee              external/effcee
 git clone --depth=1 https://github.com/google/re2                 external/re2
-curl -L http://llvm.org/svn/llvm-project/cfe/trunk/tools/clang-format/clang-format-diff.py -o utils/clang-format-diff.py;
+# The --fail flag causes the command to fail on HTTP error response codes, like 404.
+curl -L --fail https://raw.githubusercontent.com/llvm/llvm-project/main/clang/tools/clang-format/clang-format-diff.py -o utils/clang-format-diff.py
```
I don't think this is resolved yet. I would not expect to see these changes. These changes are from a commit that I did.

sliu-UIUC replied:
I'll remove those.

sliu-UIUC replied:
> I don't think this is resolved yet. I would not expect to see these changes. These changes are from a commit that I did.

I don't think I have these two lines in my `build.sh`. I don't know why you are seeing these.

sliu-UIUC replied:
Checked. Done. Use `git merge master` so there should be no conflict now.

paulthomson commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2021 Shiyu Liu
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
This copyright header looks incorrect.

paulthomson replied:
It still looks incorrect to me.

sliu-UIUC replied:
Corrected copyright header.

paulthomson commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void TransformationSwapTwoFunctions::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // Found the two functions in ir_context and swap their position.
+  // Offsets mark the relevant distance of the function from module().begin().
```
I don't think this comment is correct.

paulthomson commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void TransformationSwapTwoFunctions::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // Found the two functions in ir_context and swap their position.
+  // Offsets mark the relevant distance of the function from module().begin().
+  bool func1_found = false;
+  bool func2_found = false;
+
+  // Initialize the position (underlying: UptrVectorIterator<Function>)
+  opt::Module::iterator func1_it = ir_context->module()->begin();
+  opt::Module::iterator func2_it = ir_context->module()->begin();
+  for (auto& func : *ir_context->module()) {
+    if (func.result_id() == message_.function_id1()) func1_found = true;
+    if (func.result_id() == message_.function_id2()) func2_found = true;
+    // Once we found the target function, we stop increment iterator and thus
+    // after one iteration, func1_it and func2_it should be the iterator with
+    // their updated position.
+    // If we have not found (ie. found = false), we kept incrementing.
+    if (!func1_found) ++func1_it;
+    if (!func2_found) ++func2_it;
+  }
```
I suggest using `std::find_if` or add another utility function to fuzzer_util.cpp/h that is like `FindFunction`, but returns an iterator (pointing to end if not found). 

Then just find each function id, one after the other. Assert that they are both found.

sliu-UIUC replied:
Fixed with `GetFunctionIterator` in `fuzzerutil`.

paulthomson commented on source/opt/function.h:
```diff
@@ -54,6 +54,9 @@ class Function {
   Instruction& DefInst() { return *def_inst_; }
   const Instruction& DefInst() const { return *def_inst_; }
 
+  // Add move constructor for swap utility. 
+  Function(Function&&) = default;
+
```
I don't think this is needed now.

paulthomson commented on source/reduce/reducer.cpp:
```diff
@@ -54,10 +54,10 @@ void Reducer::SetInterestingnessFunction(
 Reducer::ReductionResultStatus Reducer::Run(
-    std::vector<uint32_t>&& binary_in, std::vector<uint32_t>* binary_out,
+    const std::vector<uint32_t>& binary_in, std::vector<uint32_t>* binary_out,
     spv_const_reducer_options options,
     spv_validator_options validator_options) {
-  std::vector<uint32_t> current_binary(std::move(binary_in));
+  std::vector<uint32_t> current_binary(binary_in);
```
Similar to my previous comment: I don't think we should see these changes. Please try squashing your changes and/or rebasing. These commits should not be in your PR.

sliu-UIUC replied:
Resolved conflicts.

paulthomson commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,248 @@
+
+  // Function should not swap with itself
+
+  // Permutation has invalid values 2
```
This comment looks incorrect.

sliu-UIUC replied:
Removed.

sliu-UIUC commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,76 @@
+    : message_(std::move(message)) {}
+
+TransformationSwapTwoFunctions::TransformationSwapTwoFunctions(uint32_t id1, uint32_t id2) {
+  message_.set_function_id1(id1);
+  message_.set_function_id2(id2);
```
> You could assert that the function ids cannot be the same here.

If I assert that id cannot be the same here, wouldn't it fail the test because `assert` will abort when it fails?

paulthomson replied:
Yes, the line that contains `TransformationSwapTwoFunctions(4, 4)` in your test would crash, but you could use ASSERT_DEATH or remove that line.

sliu-UIUC replied:
Fixed.

paulthomson reviewed (commented):
The update test looks good! There are still a few issues that need resolving before we can merge.

paulthomson commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,62 @@
+                 transformations) {}
+
+void FuzzerPassSwapFunctions::Apply() {
+  // Here we start by doing exhaustive swap testing:
+  // For every function a and function b, where a and b has a valid id in
+  // [function_ids] set and a and b are not the same function (ie, have a
+  // different id). For every combination of a and b, we do
+  // transformation_swap_two_functions(a.id, b.id) and make sure everyone of
+  // them have the correct result returned.
```
Is this comment now out of date?

sliu-UIUC replied:
Outdated. Removed.

paulthomson commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,62 @@
+              GetFuzzerContext()->GetChanceOfSwappingFunctions())) {
+        continue;
+      }
+      // We do a swap between functions and break if such swap cannot be
+      // performed.
```
Is this comment now out of date?

sliu-UIUC replied:
Outdated. Removed.

paulthomson commented on source/fuzz/protobufs/spvtoolsfuzz.proto:
```diff
@@ -2260,6 +2261,14 @@ message TransformationSwapConditionalBranchOperands {
 }
 
+message TransformationSwapTwoFunctions {
+   // A transformation that swaps the position of two functions within the same module.
+   
+   // the IDs for the two functions that are swapped.  
+    uint32 function_id1 = 1;
+    uint32 function_id2 = 2; 
```
Nit: inconsistent indentation.

sliu-UIUC replied:
Indentation corrected.

paulthomson commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,74 @@
+  auto func1_ptr = ir_context->GetFunction(message_.function_id1());
+  auto func2_ptr = ir_context->GetFunction(message_.function_id2());
+  if (!func1_ptr || !func2_ptr ||
+      func1_ptr->result_id() == func2_ptr->result_id())
```
No need to check this. You can assert that both ids are different, if you want.

sliu-UIUC replied:
Removed.

paulthomson reviewed (commented):
Nice work! I left a few remaining comments.

sliu-UIUC commented:
> Changes:
> 
> * Added `GetFunctionIteartor()` in `fuzzer_util`. Now the `apply()` in `transformation_swap_two_functions.cpp` calls `GetFunctionIteartor()`  in `fuzzerutil`.
> * Changed `transformation_swap_two_functions.cpp` and added `assert` to check either function is not found, i.e., point to `ir_context->module()->end()`.
> * Attempted clang-format with version 5.0.0.
> 
> Update:
> 
> * Resolved previous comments: Done.
> * Random files formated: Fixed. This happened due to I ran clang-format on all the files, and clang-format is not happy with having function implementations squeezed into one line,  the following files that doesn't have to do with my PR have been reversed to what it was in the master branch:
>   
>   * `source/fuzz/fuzzer_context.h`
>   * `source/opt/block_merge_pass.h`
>   * `source/opt/decoration_manager.h`
>   * `source/opt/function.h`
> * check_copyright.py: reversed the clang-format on .py file that shouldn't happen before. Added contributor.
> * Google style code comment: I think it is good for now.
> * Code comments: I didn't see much comments that immediately bothers me, but perhaps a deeper level code review would be really helpful. I will look into it a bit more later on.
> 
> Issue: #4026

sliu-UIUC commented:
> Changes:
> 
> * Added `GetFunctionIteartor()` in `fuzzer_util`. Now the `apply()` in `transformation_swap_two_functions.cpp` calls `GetFunctionIteartor()`  in `fuzzerutil`.
> * Changed `transformation_swap_two_functions.cpp` and added `assert` to check either function is not found, i.e., point to `ir_context->module()->end()`.
> * Attempted clang-format with version 5.0.0.
> 
> Update:
> 
> * Resolved previous comments: Done.
> * Random files formated: Fixed. This happened due to I ran clang-format on all the files, and clang-format is not happy with having function implementations squeezed into one line,  the following files that doesn't have to do with my PR have been reversed to what it was in the master branch:
>   
>   * `source/fuzz/fuzzer_context.h`
>   * `source/opt/block_merge_pass.h`
>   * `source/opt/decoration_manager.h`
>   * `source/opt/function.h`
> * check_copyright.py: reversed the clang-format on .py file that shouldn't happen before. Added contributor.
> * Google style code comment: I think it is good for now.
> * Code comments: I didn't see much comments that immediately bothers me, but perhaps a deeper level code review would be really helpful. I will look into it a bit more later on.
> 
> Issue: #4026



> Changes:
> 
> * Added `GetFunctionIteartor()` in `fuzzer_util`. Now the `apply()` in `transformation_swap_two_functions.cpp` calls `GetFunctionIteartor()`  in `fuzzerutil`.
> * Changed `transformation_swap_two_functions.cpp` and added `assert` to check either function is not found, i.e., point to `ir_context->module()->end()`.
> * Attempted clang-format with version 5.0.0.
> 
> Update:
> 
> * Resolved previous comments: Done.
> * Random files formated: Fixed. This happened due to I ran clang-format on all the files, and clang-format is not happy with having function implementations squeezed into one line,  the following files that doesn't have to do with my PR have been reversed to what it was in the master branch:
>   
>   * `source/fuzz/fuzzer_context.h`
>   * `source/opt/block_merge_pass.h`
>   * `source/opt/decoration_manager.h`
>   * `source/opt/function.h`
> * check_copyright.py: reversed the clang-format on .py file that shouldn't happen before. Added contributor.
> * Google style code comment: I think it is good for now.
> * Code comments: I didn't see much comments that immediately bothers me, but perhaps a deeper level code review would be really helpful. I will look into it a bit more later on.
> 
> Issue: #4026

sliu-UIUC commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,63 @@
+    // For every combination of a and b, we do transformation_swap_two_functions(a.id, b.id) and 
+    // make sure everyone of them have the correct result returned. 
+    // When the function space is large, we might choose functions arbitrarily from the function space
+    // and do random swaps.
```
Removed and edited.

sliu-UIUC commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,76 @@
+  // Debugging
+  // for(auto& func: *ir_context->module()) {
+  //   std::cout<<func.result_id()<<" "<<func.PrettyPrint()<<std::endl;
+  // }
```
Removed unnecessary comments.

afd commented on source/fuzz/fuzzer_context.cpp:
```diff
@@ -153,6 +153,7 @@ const std::pair<uint32_t, uint32_t> kChanceOfReplacingParametersWithStruct = {
 const std::pair<uint32_t, uint32_t> kChanceOfSplittingBlock = {40, 95};
 const std::pair<uint32_t, uint32_t> kChanceOfSwappingConditionalBranchOperands =
     {10, 70};
+const std::pair<uint32_t, uint32_t> kChanceOfSwappingFunctions = {10, 10};
```
This gives a lower bound and an upper bound on the probability with which functions will be swapped. 10% seems like an OK lower bound, but I suggest a higher upper bound. Let's go with 90%. So {10, 10} should become {10, 90}.

sliu-UIUC replied:
Changed to 90 upperbound.

afd commented on source/fuzz/pass_management/repeated_pass_instances.h:
```diff
@@ -167,6 +168,7 @@ class RepeatedPassInstances {
   REPEATED_PASS_INSTANCE(ReplaceParamsWithStruct);
   REPEATED_PASS_INSTANCE(SplitBlocks);
   REPEATED_PASS_INSTANCE(SwapBranchConditionalOperands);
+  REPEATED_PASS_INSTANCE(SwapFunctions);
```
Because you've made this a *final* pass (i.e., you did:

```
MaybeAddFinalPass<FuzzerPassSwapFunctions>(&final_passes_);
```

in fuzzer.cpp) it should not be added to `repeated_pass_instances`.

I.e., there should be no changes to this file.

Happy to provide you with some more context offline about the difference between repeated and final passes.

sliu-UIUC replied:
Reverted changes in `repeated_pass_instances.h`

afd commented on source/fuzz/pass_management/repeated_pass_recommender_standard.cpp:
```diff
@@ -341,6 +341,12 @@ RepeatedPassRecommenderStandard::GetFuturePassRecommendations(
     // No obvious follow-on passes
     return {};
   }
+
+  if (&pass == pass_instances_->GetSwapFunctions()) {
+    // No obvious follow-on passes
+    return {};
+  }
+
```
Again, because this is not a repeated pass you should not deal with it in this file. I.e., there should be no changes to this file.

sliu-UIUC replied:
Reverted changes in `repeated_pass_recommender_standard.cpp`

afd commented on source/fuzz/transformation_swap_two_functions.cpp:
```diff
@@ -0,0 +1,73 @@
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  auto func1_ptr = ir_context->GetFunction(message_.function_id1());
+  auto func2_ptr = ir_context->GetFunction(message_.function_id2());
+  if (!func1_ptr || !func2_ptr) return false;
```
In general, please use braces for if statements. (It's something we like to enforce in spirv-fuzz.)

But actually I would prefer you to change this to:

```
return funct1_ptr != nullptr && func2_ptr != nullptr;
```

sliu-UIUC replied:
Fixed.

afd commented on test/fuzz/transformation_swap_two_functions_test.cpp:
```diff
@@ -0,0 +1,242 @@
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  // Function should not swap with itself.
+  ASSERT_DEATH(TransformationSwapTwoFunctions(4, 4).IsApplicable(
+                   context.get(), transformation_context),
+               "The two function ids cannot be the same.");
```
You will need to enclose this code in some #ifdef to check whether the code is being compiled in debug mode. Please grep for ASSERT_DEATH in other tests and it should be obvious what to do. The reason is that ASSERT_DEATH doesn't work if you compile in release mode.

sliu-UIUC replied:
Added `#ifndef NDEBUG` to check debug mode.

afd reviewed (changes requested):
This is looking really good!

Some minor changes requested and then we should be good to merge.

afd commented on source/fuzz/fuzzer_context.cpp:
```diff
@@ -152,11 +152,12 @@ const std::pair<uint32_t, uint32_t> kChanceOfReplacingParametersWithGlobals = {
     30, 70};
 const std::pair<uint32_t, uint32_t> kChanceOfReplacingParametersWithStruct = {
     20, 40};
+const std::pair<uint32_t, uint32_t>
+        kChanceOfSwappingAnotherPairOfFunctionVariables = {30, 90};
```
We should move this so that these constants appear in alphabetical order. (Perhaps it wasn't like that already, but it should be.)

So: before kChanceOfSwappingConditionalBranchOperands, but after kChanceOfSplittingBlock.

Could you do that please?

afd replied:
Also the formatting check fails on this file, so please ensure you fix that.

Locally you can run:

./utils/check_code_format.sh

as long as clang-format version 5 is on your path.

sliu-UIUC replied:
> We should move this so that these constants appear in alphabetical order. (Perhaps it wasn't like that already, but it should be.)
> 
> So: before kChanceOfSwappingConditionalBranchOperands, but after kChanceOfSplittingBlock.
> 
> Could you do that please?

I seemed like I compared "Another" with "Block" but actually 
"splitting" goes before "swapping". Fixed.

sliu-UIUC replied:
> Also the formatting check fails on this file, so please ensure you fix that.
> 
> Locally you can run:
> 
> ./utils/check_code_format.sh
> 
> as long as clang-format version 5 is on your path.

I just clang formatted `fuzzer_context.cpp`. It should pass the test now.

Vasniktel commented on source/fuzz/fuzzer_pass_swap_functions.cpp:
```diff
@@ -0,0 +1,53 @@
+
+  size_t id_size = function_ids.size();
+  // Iterate through every combination of id i & j where i!=j.
+  for (size_t i = 0; i < id_size - 1; ++i) {
```
If SPIR-V module has no functions, the expression `id_size - 1` will overflow.

Vasniktel replied:
Although, I admit this is rather unlikely situation :)

afd replied:
@sliu-UIUC It would indeed be good to rectify this. Can you do it in a follow-up PR, @sliu-UIUC ?

sliu-UIUC replied:
> If SPIR-V module has no functions, the expression `id_size - 1` will underflow.

That's a great catch. I wonder if we use something like this: 
`
for (size_t i = 1; i < id_size ; ++i) {
    size_t tmp = i-1; 
    ...
}
`
would resolve this issue for good.

Vasniktel replied:
You can just do `for (size_t i = 0; i < id_size; ++i)` in the outer loop. Also, it's ok to replace `id_size` with `function_ids.size()` (i.e. if you want to).

sliu-UIUC replied:
> @sliu-UIUC It would indeed be good to rectify this. Can you do it in a follow-up PR, @sliu-UIUC ?

Hi sir this question might sound trivial, but is there a specific format (i.e., title, content) that I need to have for a follow up? 
I assume that the content should quote this issue and explain the problems we are addressing, and the title will be meantion its a follow-up?

Vasniktel replied:
Good question! Just describing the issue in the title (e.g. `spirv-fuzz: Fix underflow in <transformation name>`) and providing a link to the comment in the description (https://github.com/KhronosGroup/SPIRV-Tools/pull/4236#discussion_r621114293) is enough for me :)

sliu-UIUC replied:
> Good question! Just describing the issue in the title (e.g. `spirv-fuzz: Fix underflow in <transformation name>`) and providing a link to the comment in the description ([#4236 (comment)](https://github.com/KhronosGroup/SPIRV-Tools/pull/4236#discussion_r621114293)) is enough for me :)

Hi sir thanks for the feedback. I have addressed the issues you mentioned here: #4253.

Vasniktel commented on source/fuzz/fuzzer_util.cpp:
```diff
@@ -1925,6 +1925,15 @@ bool NewTerminatorPreservesDominationRules(opt::IRContext* ir_context,
   return true;
 }
 
+opt::Module::iterator GetFunctionIterator(opt::IRContext* ir_context,
+                                          uint32_t function_id) {
+  for (auto iter = ir_context->module()->begin();
+       iter != ir_context->module()->end(); ++iter) {
+    if ((*iter).result_id() == function_id) return iter;
+  }
+  return ir_context->module()->end();
+}
```
It might be better to use `std::find_if` instead of implementing it manually.

afd replied:
@sliu-UIUC This is a good idea; can you also do it in a follow-up PR?

sliu-UIUC replied:
> It might be better to use `std::find_if` instead of implementing it manually.

Hi Vasyl, I rarely used `std::find_if` so I don't know how to use it that much. But when I do try using `find_if` in the function
iterator, it seemed like the built-in `find_if` doesn't work well with `opt::module::iterator` type and kept throwing out errors.

Vasniktel replied:
On catch with find_if is that the lambda takes a value an iterator points to, not the iterator itself. In your case, it will be
```cpp
std::find_if(ir_context->module()->begin(), ir_context->module()->end(), [result_id](const opt::Function& f) {
  return f.result_id() == result_id;
})
```
See https://en.cppreference.com/w/cpp/algorithm/find.

sliu-UIUC replied:
> One catch with find_if is that the lambda takes a value an iterator points to, not the iterator itself. In your case, it will be
> 
> ```c++
> std::find_if(ir_context->module()->begin(), ir_context->module()->end(), [result_id](const opt::Function& f) {
>   return f.result_id() == result_id;
> })
> ```
> 
> See https://en.cppreference.com/w/cpp/algorithm/find.

Got it. Thanks for the feedback!

afd reviewed (approved):
Thanks for making the final changes!
